src/fsharp/FSharp.Core/string.fs

```diff
@@ -21,16 +21,52 @@ namespace Microsoft.FSharp.Core
 
         [<CompiledName("Iterate")>]
         let iter (action : (char -> unit)) (str:string) =
-            if not (String.IsNullOrEmpty str) then
-                for i = 0 to str.Length - 1 do
-                    action str.[i] 
+            let len = length str
+
+            if len = 0 then ()
+            else
+                // Unrolling gives a ~25% boost over a regular for loop on an older processor, more on a modern processor
+                let mutable i = 0
+                while i < len - len % 8 do
```

Comment:
Unrolling this is not a good look ... we should bug the jit, if they don't do a good job on this stuff.

Reply:
That's been done in 2015, and in 2017 again, with little success and it can still be years before this gets back in. And even if they would finally implement the changes, it won't help us here, as this scenario is not predictable (the from-to is variable, depending on input).

Do note that the BCL uses manual loop-unrolling in many situations, they're also not waiting for the JIT to implement this.

I would argue that this is a specific case where unrolling _does make sense_, because the generated code is so small. The measured improvement is, imo, large enough to warrant the change (note that for `String.mapi` I took the opposite stance, as the benefit is small there).

I'd also argue that this one more likely improves across the board on many CPU's, regardless of future improvements of the JIT.